Add File Size dialog and show size in attachment box

AbeJellinek opened the pull request:
Exposed as an optional item tree column and always shown in the attachment box. Getting sorting to work reliably (particularly when switching between libraries) required refactoring the item tree's caching logic slightly.

Closes #271.

AbeJellinek committed:
Show file size in item tree and attachment box

dstillman reviewed (commented):
Sorting isn't working right for me:

![file size](https://user-images.githubusercontent.com/783082/153132388-85dfc4e1-bd44-4b78-a4ac-ca52ed270d3d.png)

But I think we need to reconsider this. The attachment box is fine, but I don't think we can add this to the main items list — if people leave this on, it will drastically decrease Zotero's performance after every startup. Just turning it on makes scrolling sluggish for me and makes my fans spin up, and sorting by that column would necessitate calculating the value for every column. There's also no reason to sort this by size ascending, and no great need for filtering or for viewing by collection.

I think it'd make more sense to create a new option in Tools → Manage Attachments ("Disk Usage" or whatever) with a library drop-down (`buildLibraryMenuHTML()`) that showed a window with a new tree that showed top-level items (title/creator only?) sorted by size descending. We could perhaps show attachments underneath each one, also size descending, but I don't think that's particularly necessary. Activating (e.g., double-clicking) the row would select the item in the main window, like Advanced Search.

We could later add options for deleting or purging, but that's a little tricky — sending to the trash wouldn't change anything, and we can't do local purging at least until we finish #1919 (which has a [library-level storage breakdown](https://github.com/zotero/zotero/pull/1919#issuecomment-773430096) of its own) and can know if a file is safely on the server.

For now, the most important thing would just be letting people see the largest items in their libraries.

dstillman commented:
In the attachment box, "0 B" is a little confusing when the file isn't present. "N/A"?

In a dedicated window, we could just not show items without files or with missing files.

AbeJellinek committed:
Revert item tree changes

AbeJellinek committed:
Improve display in attachment box

dstillman commented on chrome/content/zotero/bindings/attachmentbox.xml:
```diff
@@ -302,6 +309,19 @@
+						});
+					}
+					else {
+						fileNameRow.hidden = true;
```
`fileSizeRow`

dstillman commented:
I don't think it was necessary to take out the caching. This way, "Loading…" flickers every time you select an attachment. Instead, we should cache the value, display it immediately if present, and get asynchronously after that (updating the cached/display value if changed). Again, similar to Get Info on a folder — the first time it shows "Calculating", but after that it shows the cached value immediately and updates a moment later if the value has changed.

dstillman commented:
(And we'll want the caching anyway when we display this in a separate window, so that you don't have to wait for it to scan all files if you reopen the window.)

AbeJellinek committed:
Cache file sizes again

AbeJellinek commented:
Re-added caching of the raw values. There's still some flicker, though, because the `Zotero.Item` object doesn't actually seem to be the same each time the attachment box is loaded. It'll be the same with a separate disk usage window - populating the list takes a call to `Zotero.Items.getAll()`, which returns fresh `Zotero.Item` objects. Is there a non-hacky way to persist cached values between loads without touching the database?

AbeJellinek committed:
Add basic disk usage dialog

Extracted item icon logic from itemTree.jsx.

AbeJellinek committed:
Remove unused DTD

dstillman commented on chrome/locale/en-US/zotero/zotero.properties:
```diff
@@ -416,6 +419,11 @@ pane.item.attachments.select			= Select a File
+pane.item.attachments.sizeSuffix.bytes			= B
+pane.item.attachments.sizeSuffix.kilobytes			= KB
+pane.item.attachments.sizeSuffix.megabytes			= MB
+pane.item.attachments.sizeSuffix.gigabytes			= GB
```
No longer needed, right?

dstillman commented on chrome/content/zotero/diskUsage.jsx:
```diff
@@ -0,0 +1,172 @@
+			div.innerHTML = "";
+		}
+		else {
+			div = document.createElementNS("http://www.w3.org/1999/xhtml", 'div');
```
Move to `const HTML_NS` up top and use for all

dstillman reviewed (changes requested):
Nice!

A few things:

- We need some sort of feedback in the list while sizes are being calculated. Right now it's impossible to know anything is happening. We could just do a determinate progress bar (since we know the number of parent items) in the center, or we could update the list progressively (and maybe also show a progress bar or spinner somewhere in that case?).
- We need some sort of indication when calculations are done but there's nothing to show in a given library.
- Activating rows doesn't select items in the main window for me. It just brings the main window forward.

For the attachment box, what about this?

> In the attachment box, "0 B" is a little confusing when the file isn't present. "N/A"?

I'll look into the "Loading…"/Zotero.Item thing.

Are we currently caching file sizes as long as the app is open? That might be fine, but what happens if you delete an attachment on disk? Seems like we should at least check for that (even if we start by showing the cached value in the various places).

AbeJellinek committed:
Indicate loading progress and empty collection

AbeJellinek committed:
Fix item activation

AbeJellinek committed:
Show message if file is gone

AbeJellinek committed:
Recalculate sizes after first showing cached view

Instead of updating with new items as they're loaded, the item list now
shows as "Loading..." until it's ready. This is much faster - the many
asynchronous setState() calls were hanging the UI.

qqobb commented:
I have two comments regarding `formatFileSize`.

1. In Windows Explorer, I'm getting the file size column in KB for all files. This seems like a reasonable choice: (i) Sorting by file size will just sort numbers. (ii) Changing units could be distracting. (iii) It's easy to spot large files. Perhaps you could consider a similar approach in the item tree.

```
 6,105 KB
    12 KB
     1 KB
```

Assuming that most files are in the MB range, MB could be the common unit:

```
 12.55 MB
  1.52 MB
  0.13 MB
```

Maybe use `formatFileSize(size, format)` for that?

2. In the file properties, accessed through the context menu, Windows shows both a short and a long format, e.g., `1.82 MB (1,915,981 bytes)`. It seems they're using 1024 based units for [Megabyte](https://en.wikipedia.org/wiki/Megabyte). In case you want to support that, you could use the `formatSize` function here: https://stackoverflow.com/a/40879101. Your current approach seems to follow SI units, which should be ok, but I'm not sure how common this is.

AbeJellinek commented:
>In Windows Explorer, I'm getting the file size column in KB for all files. This seems like a reasonable choice: (i) Sorting by file size will just sort numbers. (ii) Changing units could be distracting. (iii) It's easy to spot large files. Perhaps you could consider a similar approach in the item tree.

(i): This doesn't really gain us anything, since we always have the size in bytes and only format before displaying to the user. Sorting uses the numerical size, not lexical order (not useful for numbers). (ii): That's fair, but sizes are now displayed in a separate "Disk Usage" dialog that sorts by size descending, so "GB" items will always be above "MB" items will always be above "KB" items. Pretty easy to parse visually. (iii): That's fair, but I think the always-sorted Disk Usage dialog makes spotting large items easy enough.

>Assuming that most files are in the MB range, MB could be the common unit

Definitely worth considering.

>In the file properties, accessed through the context menu, Windows shows both a short and a long format ... It seems they're using 1024 based units for [Megabyte](https://en.wikipedia.org/wiki/Megabyte).

I was modeling my approach off the macOS Finder, which uses SI prefixes:

<img width="220" alt="image" src="https://user-images.githubusercontent.com/1770299/155202258-d185e94c-2740-4232-8af7-caf1746d8c2f.png">

I guess we could make the prefix definitions depend on the user's platform.

AbeJellinek committed:
Use 1024-based units on Windows

AbeJellinek commented:
OK, I think this is ready.

AbeJellinek committed:
Fix Scaffold load error when fontSize pref is set

AbeJellinek committed:
Scaffold templates: Pull URL from doc

Link URLs could lead to redirects.

dstillman committed:
Try to open external PDF reader to page number for "Show on Page"

PDF-handling code from `ZoteroPane.viewAttachment()` predates
`Zotero.OpenPDF` and didn't handle page numbers for external PDF
readers, so when called from "Show on Page" in a note (via
`ZoteroPane.viewPDF()`) it just launched the file itself in the
specified handler or via the OS.

For now, change to use `Zotero.OpenPDF.openToPage()` when there's a page
number. Later we should use `Zotero.OpenPDF` for everything, but that
code is more complicated than just using `launchFile()`, so we'll need
to do some testing (particularly on Linux) to see what's best.

(Note that "Show on Page" with an external reader doesn't make a ton of
sense, since you can't actually see the annotation, but opening the
right page is better than nothing.)

Fixes #2612

dstillman committed:
Update translators, styles, and CSL locales

dstillman committed:
Update locales from Transifex

mrtcode committed:
Avoid error interrupting annotation extraction from multiple items

dstillman committed:
Update note-editor submodule

dstillman committed:
Update version

dstillman committed:
Feed import: Don't fail on OPML entry with no `title` or `text`

https://forums.zotero.org/discussion/96841/impossible-dimporter-ompl-rss

dstillman committed:
Add prefs for timeouts for downloading PDF via browser

1 second isn't enough for some users, so try 1.5, and allow for easier
debugging via a pref.

https://forums.zotero.org/discussion/96431/i-have-access-to-pdf-but-zotero-cannot-automatically-add-it
https://forums.zotero.org/discussion/97277/unable-to-download-papers-using-zotero-chrome-extension

dstillman committed:
Update note-editor submodule

dstillman committed:
citeproc-js: Add space after left margin in plain-text mode

This adds a space after the the number in some numeric styles ("[1]" in
IEEE style, "1." in Nature) when pasting bibliography entries into a
plain-text editor.

Some styles (ACS, AMA, Vancouver) will need to be updated to remove
their hard-coded spaces to avoid double spaces after this change.

Patch from @zepinglee

Fixes #2633

dstillman committed:
Update locales from Transifex

dstillman committed:
Update global schema

Includes the Preprint archiveID/number mapping for #2481

dstillman committed:
Fix support tests after 28ed3e34b

dstillman committed:
Fix detection of some emoji set as colored tags

Emoji such as ⭐️ that began as text-based Unicode characters and depend
on Variation Selector-16 to display the emoji presentation weren't being
properly detected.

mrtcode committed:
Fix library tab tooltip always staying on top

adomasven committed:
Fix cookieSandbox not overriding user agent if no cookies for domain.

Potential fix for #2635

adomasven committed:
Improve downloadPDFViaBrowser handlers and logging

adomasven committed:
Fix a MIMETypeHandler.removeHandler() bug

adomasven committed:
Log full hidden browser count after its creation

mrtcode committed:
Add support for PDF page deletion and rotation (#2595)

* Add support for PDF page deletion and rotation
Fixes #2561

dstillman committed:
Update locales from Transifex and merge new English strings

AbeJellinek committed:
Fix sorting by attachment (#2586)

dstillman committed:
Update styles submodule

dstillman committed:
Fix Emoji detection

And move to Zotero.Utilities.Internal.isOnlyEmoji()

Fixes #2643

dstillman committed:
Properly strip ZWJ in `isOnlyEmoji()`

dstillman committed:
Fix parsing of Chinese and other non-ASCII month names

https://forums.zotero.org/discussion/97683/bug-locale-issues-concerning-zotero-and-zotero-connector

dstillman committed:
Use String.fromCharCode() to load Chinese characters in strToDate() test

Follow-up to dc045250a6

AbeJellinek committed:
Fix error if New Collection prompt is canceled

adomasven committed:
Fix force-click during drag sometimes causing collections to switch

Closes #2652

adomasven committed:
Fix alt-arrow not working on macOS when editing collection names.

Fixes #2651

adomasven committed:
Limit the force-click during drag fix to macOS (4fb0661b)

mrtcode committed:
Update pdf-reader and note-editor, use specific prebuilds, tweak code

dstillman committed:
Update translators

dstillman committed:
Update locales from Transifex

dstillman committed:
Update utilities submodule and date-parsing tests

Includes strToDate() fixes

dstillman committed:
Update version

mrtcode committed:
Implement smart notes switching in contextPane (#2671)

Fixes #2650

dstillman committed:
Update pdf-reader submodule

mrtcode committed:
Show parent item title for child notes in context pane
Fixes #2681

mrtcode committed:
Adjust page rotation labels in PDF reader
Fixes #2669

dstillman committed:
Update global schema and utilities for CSL 1.0.2

mrtcode committed:
Fix disabled "Add to Note" when a child note is open
Fixes #2687

tnajdek committed:
Don't try to remove temporary dir after import

Previously we had a structure that also created folders in temp dir,
that has since changed but code in cleanup phase to remove containing
dir has been accidentally left unremoved.

tnajdek committed:
Mendeley import: detect win close, better feedback

* Detect wizard cancel/close and interrupt import. This will still
  have to wait for current fetch (file or metadata) to complete but will
  then advance to the cleanup stage
* Advance progress bar during metadata fetch
* Add some extra logging

tnajdek committed:
Fix a bug in regex extracting fields to "extra"

Because regex is built using a template string, \s* is actually escaped
into s*, i.e. literal "s" appearing 0 or more times. In most cases this
would mean that output can have spacing slightly off. In extreme case,
when identifier starts with letter "s", this could this could lead to
identifier being stored incorrectly.

Also adjusted tests to be more strict and mock data to cover this case.

tnajdek committed:
Lint code related to Mendeley importer

tnajdek committed:
Improve log output for multiple Mendeley imports

Helps with debugging cases where new import is started while previous
one is still being cancelled

dstillman committed:
Validate fields in ItemTree::getSortFields()

To avoid startup hang if a plugin does something bad:

https://forums.zotero.org/discussion/comment/411843/#Comment_411843

Fixes #2692

dstillman committed:
Fix test failure after a8ed30ce8

We added a `toggleSort()` call on the title, so then when called again
it reverses the direction. To avoid, sort by creator and then title in
the second test.

dstillman committed:
Add CSL 1.0.2 locators and switch to CSL locales for localization

And move Page to middle of list, while still selecting by default

https://forums.zotero.org/discussion/comment/396602/#Comment_396602

dstillman committed:
Update locales from Transifex

dstillman committed:
Update pdf-reader and pdf-worker submodules

AbeJellinek committed:
`recursiveCollections`: Remove from all subcollections on delete (#2689)

dstillman committed:
Merge pull request #2693 from tnajdek/mendeley-import-improve-feedback

Multiple improvements to Mendeley importer

dstillman committed:
Tweak confirmation message for recursive collection-item removal

I guess the "its" is technically a little clearer, but it's a bit
awkward, and I think it's reasonably implied.

dstillman committed:
Merge new English strings

dstillman committed:
Add localized string for `timestamp` CSL locator

Not included in the CSL locales:

https://forums.zotero.org/discussion/comment/412060/#Comment_412060

dstillman committed:
Merge new English strings

dstillman committed:
CookieSandbox: Fix error setting request cookies

Regression from 6af3b0276a, showing up as "trackedBy is undefined" error

https://forums.zotero.org/discussion/98322/no-snapshot-or-pdf-autosaved-using-zotero-connector-on-safari

dstillman committed:
Increase `downloadPDFViaBrowser.onLoadTimeout` to 3000

Apparently necessary for some people to get PDFs on ScienceDirect. We
should come up with a better way of detecting the page load if we can,
so that we can avoid this long of a timeout for people who just will
never have access to ScienceDirect PDFs.

https://forums.zotero.org/discussion/comment/411056/#Comment_411056

Addresses #2698

dstillman committed:
Fix sorting by Item Type and feed item itemID

Regression from a8ed30ce803

dstillman committed:
Fix sorting of attachments and notes columns

Regression from a8ed30ce803

dstillman committed:
Sort empty rows last for Date/Year columns when descending

Date columns start in descending order, so put empty rows at end

Also remove a seemingly unnecessary exception for Title when not the
primary sort and just always sort it with empty rows first

dstillman committed:
Revert "Validate fields in ItemTree::getSortFields()"

This reverts commit a8ed30ce803a9c5b3b5c04d50f6bd6b94faeee86 and related
commits.

We'll address breakage from invalid sort fields another way, without
inconveniencing plugin authors.

https://groups.google.com/g/zotero-dev/c/kc0-C6-SA74/m/bhHniGceAQAJ

dstillman committed:
Better fix for errors from invalid sort fields

Just catch the error from `ItemTree::sort()` and clear the
secondary-sort and fallback-sort prefs so that sorting works on the next
attempt.

Replacement for a8ed30ce803a9c5b3b5c04d50f6bd6b94faeee86

https://groups.google.com/g/zotero-dev/c/kc0-C6-SA74/m/bhHniGceAQAJ

dstillman committed:
Fix test failure after 26a5e2900a

mrtcode committed:
Include quotation marks for highlights by default, but allow to disable
Fixes #2697

mrtcode committed:
Don't include quotes when {{highlight}} is placed in a blockquote
Fixes #2697

mrtcode committed:
Prevent delayed context pane expansion and pdf-reader rerender/resize when switching from library tab

AbeJellinek committed:
Prevent deleting My Publications

Previously showed a blank confirmation dialog and logged an error.

mrtcode committed:
Use Quick Copy (Markdown) for direct plain-text copying from note-editor
Fixes #2505

dstillman committed:
Add warning in toolbar for 'dev' channel builds

dstillman committed:
Apply Big Sur styling when macOS version comes through correctly

Follow-up to zotero/zotero-standalone-build@f7a506a32 -- now that some
people are somehow getting an accurate macOS version number, we need to
apply the Big Sur CSS to all higher versions.

https://forums.zotero.org/discussion/98406/zotero-will-not-open-mac-os-monterey

dstillman committed:
Update pdf-reader submodule

dstillman committed:
Increase Mendeley API timeout to 60 seconds

To see if we can fix the 30-second timeout here:

https://forums.zotero.org/discussion/comment/412750/#Comment_412750

dstillman committed:
Disable sort tests from 26a5e2900a for now

Until I can figure out the race-condition breakage they trigger later in
the file

dstillman committed:
Update locales from Transifex

dstillman committed:
Update submodules

dstillman committed:
Update translate submodule

dstillman committed:
Update note-editor submodule

dstillman committed:
Update version

dstillman committed:
Localize "Export PDF…" menu option

Fixes #2710

adomasven committed:
Update SingleFile submodule

adomasven committed:
Changes for updated SingleFile submodule

adomasven committed:
Fixes some assets not being saved with SingleFile snapshots

Part of zotero/zotero-connectors#394

When saving an item (e.g. on arxiv) with a site translator, the
translator may add a custom snapshot attachment URL which does not match
the translation URL (the URL in the browser where translation was
initiated) in which case Zotero handles snapshot saving. Zotero loads
the page in a background browser. By default images are not loaded in
this browser. SingleFile due to CORS restrictions has to refetch all
resources that are not from the same domain or subdomain, but uses
pre-loaded resources (images) from the same domain. Before this change
any images that were from the same domain as the saved website were not
saved in the snapshot. This commit fixes that.

mrtcode committed:
Fix regression that prevents showing new notes in contextPane notes list

dstillman committed:
Update global schema

dstillman committed:
Update version

mrtcode committed:
Fix regression that prevents showing the last selected tab on startup
Fixes #2723

mrtcode committed:
Freeze note schema version to 8
We can do it because it was only used to create a note from annotations.

No need to update schema version in Zotero client, unless using new
features when creating a note from annotations.

adomasven committed:
Do not load-timeout on browser PDF download during content load

Closes #2726

AbeJellinek committed:
Merge branch 'master' into file-size

AbeJellinek committed:
Fix Title column label

gutjuri commented:
When will this be merged?

LSartore commented:
Hi, what is the status of this pull request ?
